### PR TITLE
Fix/limit ranges

### LINKED ISSUE #119: AnalysisCriterion start/finish indices
```
TradingRecord record = manager.run(strategy, startWindow, endWindow);
double improvement = new VersusBuyAndHoldCriterion(
     new TotalProfitCriterion()).calculate(series, record);
```
"improvement" will be the record (startWindow to endWindow) compared against buy-and-hold for the entire series.  This is not what I wanted so I added
```
public double calculate(TimeSeries series, TradingRecord tradingRecord,
     int startIndex, int finishIndex) {
        TradingRecord fakeRecord = new BaseTradingRecord();
        fakeRecord.enter(startIndex);
        fakeRecord.exit(finishIndex);
        return criterion.calculate(series, tradingRecord) / criterion.calculate(series, fakeRecord);
    }
```
and changed the code to
```
TradingRecord record = manager.run(strategy, startWindow, endWindow);
double improvement = new VersusBuyAndHoldCriterion(
     new TotalProfitCriterion()).calculate(series, record, startWindow, endWindow);
```
Not sure if this should belong here but if it's put in the AnalysisCriterion interface then all the implementing classes will need the addition.  Happy to submit a PR for this simpler change.

### PR DESCRIPTION
**Pull request by edlins**
Fixes #119 .

Changes proposed in this pull request:
- Limit [Versus]BuyAndHold calculations to range of trading record
- Fix test cases with no trades to return 1d
- 

- [x] added an entry to the unreleased section of `CHANGES.md`

### CODE REVIEW

**Comment by edlins** [comment]
This solves #119  in a very simple manner.  The only contentious changes are to the test cases where the trading record has no closed trades.  Per Marc in #125 , trades that aren't closed should not affect the score of a strategy so this should be a correct fix.

File: ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/BuyAndHoldCriterion.java [hunk 1]
```diff
@@ -35,7 +35,13 @@ this software and associated documentation files (the "Software"), to deal in
...
     @Override
     public double calculate(TimeSeries series, TradingRecord tradingRecord) {
-        return series.getBar(series.getEndIndex()).getClosePrice().dividedBy(series.getBar(series.getBeginIndex()).getClosePrice()).doubleValue();
+        if (tradingRecord.getTradeCount() == 0) {
+            return 1.0d;
+        }
+        int firstEntryIndex = tradingRecord.getTrades().get(0).getEntry().getIndex();
+        int lastExitIndex = tradingRecord.getTrades().get(tradingRecord.getTradeCount()-1).getExit().getIndex();
+
+        return series.getBar(lastExitIndex).getClosePrice().dividedBy(series.getBar(firstEntryIndex).getClosePrice()).doubleValue();
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/BuyAndHoldCriterion.java:44]: Is it correct to say that the range of the `TradingRecord` is just from first entry to last exit index of its trades? Maybe the run was over a much bigger range, but the first trade started after some bars or the last exit was earlier. In my opinion in most cases this assumption is not valid, the `TradingRecord` needs start and end indices that are set in TimeSeries#run()

**Comment by edlins** [comment]
You are correct (again.)  I think I've got the correct solution and I'm spending extra time developing robust test cases.  I'll push another commit as soon as I can.

**Comment by edlins** [comment]
This is more complicated than I thought.  When you introduce indices, you are introducing a context against which everything is evaluated.  But TradingRecord has a fixed context based on the chronological sequence of adding trades.  The whole concept of a current trade depends on the context, but TradingRecord#getCurrentTrade() does not account for an arbitrary context.  So, when looking at different subranges of a TradingRecord, getCurrentTrade() will always return the same Trade even if it's outside of the subrange.  For example, you have a TradingRecord with a Trade that buys at index 0 and sells at index 10.  Now you calculate from index 0 to index 5.  The calculation views currentTrade.isNew() as true, isOpened() as false, and isClosed() as false.  It should see !isNew(context), isOpened(context), and !isClosed(context).  All of the context-dependent methods will need the context indices.  This would take a pretty serious re-write of TradingRecord and BaseTradingRecord.

**Comment by edlins** [comment]
This is not ready for merging so I will close it until it's ready.  I'll keep the branch updated in my fork if you want to follow along.

**Pull request closed without merging**